geomesa-convert/geomesa-convert-common/src/main/scala/org/locationtech/geomesa/convert/SimpleFeatureValidator.scala

```diff
@@ -8,168 +8,256 @@
   }
 
-  object HasDtgValidator {
-    val name: String = "has-dtg"
+  def unapplySeq(validator: SimpleFeatureValidator): Option[Seq[String]] = {
+    validator match {
+      case f: FeatureValidator   => Some(f.names())
+      case f: CompositeValidator => Some(f.names())
+      case _ => None
+    }
   }
 
-  class HasDtgValidator extends SimpleFeatureValidator {
-    import org.locationtech.geomesa.utils.geotools.RichSimpleFeatureType.RichSimpleFeatureType
-
-    private var i: Int = -1
-    private var lastErr: String = _
+  /**
+    * Factory for validators
+    */
+  trait ValidatorFactory {
+    def name: String
+    def validator(sft: SimpleFeatureType, config: Option[String]): Validator
+  }
 
-    override def name: String = HasDtgValidator.name
+  /**
+    * Simplified validator interface for delegating
+    */
+  trait Validator {
+
+    /**
+      * Validates a feature
+      *
+      * @param sf feature
+      * @return null if valid, otherwise error message
+      */
+    def validate(sf: SimpleFeature): String
+  }
 
-    override def init(sft: SimpleFeatureType): Unit = {
-      lastErr = null
-      i = sft.getDtgIndex.getOrElse(-1)
-    }
+  /**
+    * Simple feature validator implementation that delegates to a validator factory
+    *
+    * @param factory factory
+    */
+  class FeatureValidator(factory: ValidatorFactory, config: Option[String]) extends SimpleFeatureValidator {
+    private var validator: Validator = _
+    def names(): Seq[String] = Seq(factory.name + config.map(c => s"($c)").getOrElse(""))
+    override def init(sft: SimpleFeatureType): Unit = validator = factory.validator(sft, config)
+    override def validate(sf: SimpleFeature): String = validator.validate(sf)
+  }
 
-    override def validate(sf: SimpleFeature): Boolean = {
-      if (i == -1) { true } else {
-        val res = sf.getAttribute(i) != null
-        lastErr = if (res) { null } else { "Null date attribute" }
-        res
+  /**
+    * Evaluates multiple validators
+    *
+    * @param factories factories
+    */
+  class CompositeValidator(factories: Seq[(ValidatorFactory, Option[String])]) extends SimpleFeatureValidator {
+    private var validators: Seq[Validator] = _
+    def names(): Seq[String] =
+      factories.map { case (factory, config) => factory.name + config.map(c => s"($c)").getOrElse("") }
+    override def init(sft: SimpleFeatureType): Unit =
+      validators = factories.map { case (factory, config) => factory.validator(sft, config) }
+    override def validate(sf: SimpleFeature): String =
+      validators.foldLeft(null: String) { (error, v) =>
+        val valid = v.validate(sf)
+        if (error == null) { valid } else if (valid == null ) { error } else { s"$error, $valid" }
       }
-    }
-
-    override def lastError: String = lastErr
   }
 
-  object HasGeoValidator {
-    val name: String = "has-geo"
+  /**
+    * Validator based on the indices used by the feature type. Currently only the z2 and z3 indices have
```

Comment:
When you say 'z2' and 'z3', does that include the XZn indices?

Reply:
yeah, i forgot to update that comment but it works with XZ as well